File: credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

// validate a credit card, take the second last number and the numbers alternating before that and multiply it by two. if the number> 10 then add the digits within that number together. add it to the remaining numbers and if it ends in a zero, it is a credit card.
// to get the last digit, do % 10 and itll give you. to get the second last digit, divide by 100 then take the % of 10. this is process. i.e. loop

int main(void)
{
    // user input. must be so that a credit card number is input and not letters
    long credit_card_number;

    do
    {
        credit_card_number = get_long("Input your credit card number: ");
    }
    while (credit_card_number <= 0);

    // luhns algorith

    // this is for calculations later on
    long CC_number;
    CC_number = credit_card_number;

    // add the last number and the alternating numbers together

    long last_number, alternate_even, sum, division;
    long limit_1 = 1000000000000000;
    long sum_even = 0;

    last_number = credit_card_number % 10;

    for (division = 100; division <= limit_1; division = division * 100)
    {
        alternate_even = (credit_card_number / division) % 10;
        sum_even = sum_even + alternate_even;
    }
    sum = sum_even + last_number;

    // find the second last number and the alternating numbers and multiply by 2. if multiplication > 9 then add two digits.

    long alternate_odd, division2, bigger_than_9, sum_odd;
    long sum_odd_1 = 0;
    long sum_odd_2 = 0;

    for (division2 = 10; division2 <= limit_1; division2 = division2 * 100)
    {
        alternate_odd = ((credit_card_number / division2) % 10) * 2;

        if (alternate_odd > 9)
        {
            bigger_than_9 = (alternate_odd % 10) + (alternate_odd / 10);
            sum_odd_1 = sum_odd_1 + bigger_than_9;
        }

        else
        {
            sum_odd_2 = sum_odd_2 + alternate_odd;
        }
    }
    sum_odd = sum_odd_1 + sum_odd_2;

    // add the two results together. if number ends in 20 it is valid.

    long total_sum;
    total_sum = sum_odd + sum;

    // print whether it is AMEX, MASTERCARD, VISA or INVALID

    // finds number of digits in card

    int count = 0;

    while (credit_card_number > 0)
    {
        credit_card_number = credit_card_number / 10;
        count = count + 1;
    }

    // finds first two digits

    long digits, power;
    power = pow(10, (count - 2));
    digits = CC_number / power;

    // finds first digit

    long digit;
    power = pow(10, (count - 1));
    digit = CC_number / power;

    //American express - 15 digits, starts with 34 or 37
    // Mastercard - 16 digits, starts with 51 - 55
    // VISA - 13 or 16 digits starts with 4

    if (total_sum % 10 == 0)
    {
        if (count == 15 && (digits == 34 || digits == 37))
        {
            printf("AMEX\n");
            return 0;
        }

        else if ((count == 13 || count == 16) && digit == 4)
        {
            printf("VISA\n");
            return 0;
        }
        else if (count == 16 && (digits == 51 || digits == 52 || digits == 53 || digits == 54 || digits == 55))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    else
    {
        printf("INVALID\n");
        return 0;
    }
}

File: credit.py
# User input credit card number. Must be 16 digits long and numbers
while True:
    try:
        userInput=int(input("Input your credit card number: "))
        if userInput <=0:
            ValueError
            print("Has to be greater than zero")
            continue
    except ValueError:
        print("Enter a valid number please.")
        continue
    else:
        break

# Luhns algorithm: Disclaimer, do not take my even and odd as good variable names
# I should have really kept it as last digit and second to last digit.

# Sorting the odd numbers in CC ----------------------

# Takes user input an puts it into an array which I can then slice
cc_number_list = [int(i) for i in str(userInput)]

# Takes the list and only use the second last number and evey other number i.e. odd from the second last
odd_numbers = cc_number_list[-2::-2]

# Multiply the numbers by 2
x2_odd_numbers = [element * 2 for element in odd_numbers]

sum_of_greater = 0
sum_of_less = 0

# If odd numbers are greater than 9 then add the two integers
for i in range(len(x2_odd_numbers)):
    if x2_odd_numbers[i] > 9:
        greater_than_9_sorted = (int(x2_odd_numbers[i]/10) + x2_odd_numbers[i]%10)
        sum_of_greater = sum_of_greater + greater_than_9_sorted
    else:
        sum_of_less = sum_of_less + x2_odd_numbers[i]

# Sum the odd numbers after sorting the integers
sum_of_odd = sum_of_greater + sum_of_less

# Sorting the even numbers -----------------------------

# Takes the list and only use the last number and evey other number i.e. even from the last
even_numbers = cc_number_list[::-2]
even_sum = sum(even_numbers)

# Sum all calculations --------------------------------

final_sum = even_sum + sum_of_odd

# Determining the CC type -----------------------------

for i in range(len(cc_number_list[0:2:1])):
    #AMEX
    if (cc_number_list[0] == 3 and (cc_number_list[1] == 4 or cc_number_list[1] == 7)):
        if (len(str(userInput)) == 15):
            print("AMEX")
            break

    # Mastercard
    elif (cc_number_list[0] == 5 and (cc_number_list[1] == 1 or cc_number_list[1] == 2 or cc_number_list[1] == 3 or cc_number_list[1] == 4 or cc_number_list[1] == 5)):
        if (len(str(userInput)) == 16):
            print("MASTERCARD")
            break

    # VISA
    elif (cc_number_list[0] == 4):
        if (len(str(userInput)) == 13 or len(str(userInput)) == 16):
            print("VISA")
            break

    else:
        print("INVALID")
        break
